src/contract.ts

```diff
@@ -1,376 +1,479 @@
-   */
-  public async removeEventListener(listener: number): Promise<void> {
-    return await this.program.removeEventListener(listener);
-  }
-
-  /**
-   * Parse event `logs` for any events emitted
-   *
-   * @param logs  Array of log strings
-   * @returns     Decoded event data
-   */
-  public parseLogsEvents(logs: string[]): LogDescription[] {
-    const events: LogDescription[] = [];
-
-    for (const log of logs) {
-      const eventData = parseLogTopic(log);
-      if (eventData) {
-        const event = this.abi.parseLog(eventData);
-        events.push(event);
-      }
+    /** @internal */
+    protected buildCall(fragment: FunctionFragment, returnResult: boolean): ContractFunction {
+        return (...args: any[]) => {
+            const options = args[args.length - 1];
+            if (args.length > fragment.inputs.length && typeof options === 'object') {
+                return this.call(fragment, returnResult, args.slice(0, fragment.inputs.length), options);
+            } else {
+                return this.call(fragment, returnResult, args);
+            }
+        };
     }
 
-    return events;
-  }
+    /** @internal */
+    protected async call<T extends boolean>(
+        fragment: FunctionFragment,
+        returnResult: T,
+        args: readonly any[],
+        options?: ContractCallOptions
+    ): Promise<T extends true ? any : ContractFunctionResult> {
+        const payer = options?.payer || this.payer;
+        if (!payer) throw new Error('MISSING_PAYER_ACCOUNT'); // @FIXME: add error types
+
+        const {
+            accounts = [],
+            writableAccounts = [],
+            programDerivedAddresses = [],
+            signers = [],
+            sender = payer.publicKey,
+            value = 0,
+            simulate = false,
+            confirmOptions = {
+                commitment: 'confirmed',
+                skipPreflight: false,
+                preflightCommitment: 'processed',
+            },
+        } = options ?? {};
+
+        const seeds = programDerivedAddresses.map(({ seed }) => seed);
+        const input = this.interface.encodeFunctionData(fragment, args);
+
+        const data = Buffer.concat([
+            this.storage.toBuffer(), //                     storage @FIXME: these comments are kind of useless
+            sender.toBuffer(), //                           sender  @FIXME: better to explain why, not what
+            Buffer.from(numToPaddedHex(value), 'hex'), //   value
+            Buffer.from('00000000', 'hex'), //              hash
+            encodeSeeds(seeds), //                          seeds
+            Buffer.from(input.replace('0x', ''), 'hex'), // input
+        ]);
+
+        const keys = [
+            // @FIXME: should all these PDAs really be writable?
+            ...programDerivedAddresses.map(({ address }) => ({
+                pubkey: address,
+                isSigner: false,
+                isWritable: true,
+            })),
```

Comment:
Are these PDAs always storage accounts, is that why they are always writable? If so, it might be better to name them something else like `storageAccounts` so devs don't pass other PDAs here in the wrong order.

Reply:
So at the moment, PDAs are only used for contract creation. Any other use cases need some more attention.